### PR TITLE
FLUX LoRA Support

### PR DESCRIPTION
**Pull request by RyanJDick**
## Summary

This PR adds support for FLUX LoRA models on both quantized and non-quantized base models.

Supported formats:
- diffusers
- kohya

Full changelist:
- Consolidated LoRA handling code in `invokeai/backend/lora`
- Add support for FLUX kohya and FLUX diffusers LoRA model loading
- Add ability to either patch LoRAs or run as a sidecar model (the latter enables LoRAs to be applied to a wide range of quantized models).

## QA Instructions

Note to reviewers: I tested everything in this checklist. Feel free to re-verify any of this, but also test any LoRAs that you have. There are many small LoRA format variations, and there's a risk of breaking one of them with this change.

FLUX LoRA
- [x] Import / probe of kohya FLUX LoRA (https://civitai.com/models/159333/pokemon-trainer-sprite-pixelart?modelVersionId=779247)
- [x] Import / probe of Diffusers FLUX LoRA (https://civitai.com/models/200255/hands-xl-sd-15-flux1-dev?modelVersionId=781855)
- [x] kohya with non-quantized base model
- [x] kohya with quantized base model (should roughly match the non-quantized case)
- [x] diffusers with non-quantized base model
- [x] diffusers with quantized base model (should roughly match the non-quantized case)
- [x] Sidecar LoRA patching speed (<0.1secs after model is loaded)
- [x] Stacking multiple fused LoRA models (i.e. on top on non-quantized model)
- [x] Stacking multiple sidecar LoRA models (i.e. on top of quantized model) 

Regression Tests
- [x] SD1.5 LoRA (check output, speed and memory)
- [x] SDXL LoRA (check output, speed and memory)
- [x] `USE_MODULAR_DENOISE=1` smoke test with LoRA

Test for output regression with the following LoRA formats:
  - [x] LoRA
  - [x] LoHA
  - [x] LoKr
  - [x] IA3

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [x] _Tests added / updated (if applicable)_
- [x] _Documentation added / updated (if applicable)_

### CODE REVIEW

File: invokeai/backend/model_manager/probe.py [hunk 1]
```diff
@@ -554,12 +560,21 @@ class LoRACheckpointProbe(CheckpointProbeBase):
     """Class for LoRA checkpoints."""
 
     def get_format(self) -> ModelFormat:
-        return ModelFormat("lycoris")
+        if is_state_dict_likely_in_flux_diffusers_format(self.checkpoint):
+            # TODO(ryand): This is an unusual case. In other places throughout the codebase, we treat
+            # ModelFormat.Diffusers as meaning that the model is in a directory. In this case, the model is a single
+            # file, but the weight keys are in the diffusers format.
+            return ModelFormat.Diffusers
```
Comment (RyanJDick) [inline invokeai/backend/model_manager/probe.py:567]: Question for reviewers: Any concerns with this?

**Comment by RyanJDick** [comment]
Note to reviewers: Please test any LoRAs that you have (SD or FLUX). There are many small LoRA format variations, and there's a risk of breaking one of them with this change.

File: invokeai/backend/model_manager/load/model_loaders/lora.py [hunk 2]
```diff
@@ -68,7 +68,13 @@ def _load_model(
             model = lora_model_from_sd_state_dict(state_dict=state_dict)
         elif self._model_base == BaseModelType.Flux:
             if config.format == ModelFormat.Diffusers:
-                model = lora_model_from_flux_diffusers_state_dict(state_dict=state_dict)
+                # HACK(ryand): We assume alpha=8 for diffusers PEFT format models. These models are typically
+                # distributed as a single file without the associated metadata containing the alpha value. We chose
+                # alpha=8, because this is the default value in the PEFT library:
```
Comment (bghira) [inline invokeai/backend/model_manager/load/model_loaders/lora.py:73]: lol it's alpha=dim

**Comment by bghira** [comment]
i don't see how alpha=8 would work for any peft loras that aren't also rank=8

File: invokeai/backend/model_manager/load/model_loaders/lora.py (hunk 2 above)
Comment (RyanJDick) [reply invokeai/backend/model_manager/load/model_loaders/lora.py:73]: Agreed that's probably a more common choice. I'll fix this up.

Are you going based on https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py#L1194? Or something else?

Comment (bghira) [reply invokeai/backend/model_manager/load/model_loaders/lora.py:73]: yeah we've got some stuff WIP for making alpha an encoded value in the diffusers (peft) LoRAs but it's not really widely available yet, and diffusers' assumption is alpha=dim

**Comment by bghira** [comment]
if you're going for lycoris support note that the support for this in peft is very lacking currently, with an open issue for community help request on updating the various modules

it's very easy to wrap the upstream lycoris library modules though, as @KohakuBlueleaf has added a functional API for this use case. it just wraps the Diffusers model and has `.apply_to()` and `.merge_to()` for sidecar or fused running

there's built in methods for extracting models and playing with them in other ways. i'd highly recommend switching from peft for lycoris stuff.

**Comment by KohakuBlueleaf** [comment]
> it's very easy to wrap the upstream lycoris library modules though, as @KohakuBlueleaf has added a functional API for this use case. it just wraps the Diffusers model and has `.apply_to()` and `.merge_to()` for sidecar or fused running

`.apply_to()` is module API. Functional API is something designed for library to implement their own low-level wrapper.
(For example, `lokr_output = layer(x) + lokr_bypass_forward(x, w1, w2a, w2b ...)`)

File: invokeai/backend/model_manager/load/model_loaders/lora.py (hunk 2 above)
Comment (RyanJDick) [reply invokeai/backend/model_manager/load/model_loaders/lora.py]: Addressed in [e88d3cf](https://github.com/invoke-ai/InvokeAI/pull/6847/commits/e88d3cf2f7459c8ce1ae56e1ee72f5c2a9976fb6)

**Comment by RyanJDick** [comment]
> if you're going for lycoris support note that the support for this in peft is very lacking currently, with an open issue for community help request on updating the various modules
> 
> it's very easy to wrap the upstream lycoris library modules though, as @KohakuBlueleaf has added a functional API for this use case. it just wraps the Diffusers model and has `.apply_to()` and `.merge_to()` for sidecar or fused running
> 
> there's built in methods for extracting models and playing with them in other ways. i'd highly recommend switching from peft for lycoris stuff.

Thanks for this context. We explored using PEFT at one point, but decided against using it because it didn't meet our requirements for high patching/unpatching speeds.

I'll leave the integration of the upstream lycoris lib for future work. For now, we have our own implementation of the lycoris layers for fused execution, and partial support for sidecar execution (and easy to extend).

File: invokeai/app/invocations/flux_lora_loader.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        description=FieldDescriptions.transformer,
+        input=Input.Connection,
+        title="FLUX Transformer",
+    )
+
+    def invoke(self, context: InvocationContext) -> FluxLoRALoaderOutput:
+        lora_key = self.lora.key
+
+        if not context.models.exists(lora_key):
+            raise ValueError(f"Unknown lora: {lora_key}!")
```
Comment (brandonrising) [inline invokeai/app/invocations/flux_lora_loader.py:40]: Perhaps this should be `UnknownModelException`

File: invokeai/app/invocations/flux_lora_loader.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        if any(lora.lora.key == lora_key for lora in self.transformer.loras):
+            raise ValueError(f'LoRA "{lora_key}" already applied to transformer.')
+
+        transformer = self.transformer.model_copy(deep=True)
+        transformer.loras.append(
```
Comment (brandonrising) [inline invokeai/app/invocations/flux_lora_loader.py:46]: Any validation we should do here to ensure the lora and transformer are compatible?

File: invokeai/app/invocations/flux_denoise.py [hunk 5]
```diff
@@ -187,9 +191,41 @@ def _run_diffusion(
...
+                exit_stack.enter_context(
+                    LoRAPatcher.apply_lora_sidecar_patches(
+                        model=transformer,
+                        patches=self._lora_iterator(context),
+                        prefix="",
+                        dtype=inference_dtype,
+                    )
+                )
+            else:
+                raise ValueError(f"Unsupported model format: {config.format}")
```
Comment (brandonrising) [inline invokeai/app/invocations/flux_denoise.py:227]: Perhaps we should add the supported formats to this error message

File: invokeai/backend/lora/lora_patcher.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+            isinstance(orig_layer, LoRASidecarModule) and isinstance(orig_layer.orig_module, torch.nn.Linear)
+        ):
+            if isinstance(lora_layer, LoRALayer):
+                return LoRALinearSidecarLayer(lora_layer=lora_layer, weight=patch_weight)
+            elif isinstance(lora_layer, ConcatenatedLoRALayer):
+                return ConcatenatedLoRALinearSidecarLayer(concatenated_lora_layer=lora_layer, weight=patch_weight)
+            else:
+                raise ValueError(f"Unsupported Linear LoRA layer type: {type(lora_layer)}")
+        else:
+            raise ValueError(f"Unsupported layer type: {type(orig_layer)}")
```
Comment (brandonrising) [inline invokeai/backend/lora/lora_patcher.py:251]: May be worth adding expected types here

File: invokeai/backend/model_manager/probe.py (hunk 1 above)
Comment (brandonrising) [reply invokeai/backend/model_manager/probe.py:567]: After looking through the flow I don't believe there are significant concerns here. LoRALoader._get_model_path does treat this format as a declaration of the model being a directory, but it doesn't seem to have any issues with it being a checkpoint file instead.

File: invokeai/app/invocations/flux_lora_loader.py (hunk 3 above)
Comment (RyanJDick) [reply invokeai/app/invocations/flux_lora_loader.py:40]: From what I can tell, `UnknownModelException` started as a custom exception intended to specifically indicate an attempt to access an unknown model in an API request. It looks like that definition has since been stretched in a couple places. I'd rather not stretch it any further given that there is no special handling of a `UnknownModelException` thrown from an invocation. What do you think?

File: invokeai/app/invocations/flux_denoise.py (hunk 5 above)
Comment (RyanJDick) [reply invokeai/app/invocations/flux_denoise.py:227]: If this exception is hit, the response should be for a developer to go and add the new model format. I.e. there's no action someone could take based on knowing the supported formats. So, I figure it's best not to maintain a list of supported formats that could drift out of sync over time.

File: invokeai/backend/lora/lora_patcher.py (hunk 6 above)
Comment (RyanJDick) [reply invokeai/backend/lora/lora_patcher.py:251]: Same comment as above: Knowing the supported layer types does not help to resolve the issue, so I figured it's better not to try and maintain a list just for logging purposes.

File: invokeai/app/invocations/flux_lora_loader.py (hunk 4 above)
Comment (brandonrising) [reply invokeai/app/invocations/flux_lora_loader.py:46]: It wouldn't require a call to the db in commercial, but we can just worry about it in the future if it proves to be an issue

**Review by brandonrising — APPROVED** [review]
